### DISCUSSION TITLE
How to export a Maven-centric POM?

### DISCUSSION THREAD

**Discussion by kriegaex** [eclipse-tycho/tycho#892] [Q&A]
### Situation

I want to refactor a few things in the Java code of a multi-module Tycho project, namely Eclipse AJDT. The thing is, I am feeling totally uncomfortable with Eclipse IDE and wish to do the refactoring in IntelliJ IDEA. Unfortunately, there is no good Tycho support in IDEA, which means that if I import the project, dependencies are not recognised correctly, not even the ones between modules in the same project. I.e. IDEA can neither compile nor help to refactor anything in that project, because dependencies expressed in OSGi manifests are futile. IDEA has otherwise great Maven support, but Tycho is not Maven, to over-simplify it for my use case.

### Question

Is there any way to either make IDEA discover Tycho/OSGi dependencies or to somehow transform the information existing in OSGi manifests into an equivalent Maven POM in order to give IDEA a chance to discover the dependencies correctly, even at the possible expense of information loss with regard to what is allowed to be read in those dependencies from an OSGi perspective? I was thinking about then temporarily overwriting the original POMs while working on the Java code, simply not staging them for commit and reverting them when I am done. This would be an awful hack, but give me the chance to use an IDE other than Eclipse.

Of course, the real solution would be Tycho support for IDEA, see e.g. [IDEA-186628](https://youtrack.jetbrains.com/issue/IDEA-186628). Being a total OSGi noob, it is beyond me to understand why the existing OSGi support in IDEA does not encompass Tycho, or if there is a way to import the project directly from the OSGi modules and ignore Maven. But then I would also lose information expressed only in Maven, such as AspectJ Maven Plugin settings giving me the necessary AspectJ facets (like natures in Eclipse) in some modules.

### Epilogue

Sorry if this request seems naive and makes you think something like: "Why doesn't the guy simply learn how to use Eclipse, if he is working on an Eclipse project?" Well, I simply do not wish to be locked into an IDE I do not feel comfortable with. For most purposes, IntelliJ IDEA Ultimate is a very good choice, so I would like to stick to it. I keep using the wrong keyboard shortcuts in both IDEs after switching, just having totten used to the other one again after a while. There is no routine when working in both, I always have to focus and remember where I am at, which interrupts my flow. I want to focus on the task at hand rather than which IDE I happen to be using just in this moment. I am also working on other, non-Eclipse projects in parallel and want to avoid switching IDEs, just because I am switching projects.

**Comment by laeubi** [comment]
> IDEA has otherwise great Maven support, but Tycho is not Maven, to over-simplify it for my use case.

This is wrong, Tycho *is* maven as you can execute the project with a vanilla `mvn` environment, but probably the maven support of idea is not that great as stated? :-)

The problem is (and m2eclipse also suffers from that) that tycho uses maven extensions, so if IDEA does not execute them in a suitable way and instead relies on the *static* maven model it won't see dependencies from the *dynamic* maven model. You can try that out by running `dependency:list` on the command line (or maybe even from within IDEA).

So I think IDEA do not need to support specifically tycho, but some kind of importing the dynamic model (even though that might be slow.

You could try the following, even though this is completely untested and unsupported:

1. make sure you are using the latest Tycho release
2. make sure your build runs fine on the command line
3. enable the [tycho-packaging:update-consumer-pom](https://www.eclipse.org/tycho/sitedocs/tycho-packaging-plugin/update-consumer-pom-mojo.html) and make sure you set `skipPomGeneration=false` (always generate a consumer pom) and `deleteOnExit=false` (keep the file) and `tychoPomFilename=pom.xml` (overwrite your original pom)

that *might* gives you something idea could potentially understand...

**Disclaimer:** The main problem will be, that most probably some or all dependencies are **not** available as maven dependencies but P2 dependencies and is is unlikely IDEA will understand them! There is a ticket for that #590 but that's more a long running idea (If this crucial to your business and likes to speed up the development in that area a [sponsoring](https://github.com/sponsors/laeubi?frequency=one-time) would allow me to assign more time-slots particular issue) and I don't know how good IDEA can handle custom repositories.

**Reply by kriegaex** [reply]
OK, blame it on IDEA and m2e then, because they don't import the dynamic Maven model. Fine. That the dynamic model is necessary in order to resolve all Tycho dependencies in the first place and that IDEA might not understand P2 repositories, probably does not make it a Tycho integration problem either, but IDEA's Maven support is just bad.

As for your hints and pointers, I am really grateful and might look into them, if what I just discovered two hours after asking this question proves insufficient.

I don't know why I never found it before (wrong search terms probably), but the [Eclipse PDE Partial](https://plugins.jetbrains.com/plugin/16761-eclipse-pde-partial) IDEA plugin seems to help me resolve the dependencies. At least, after installing the plugin and re-opening the project, IDEA immediately recognised the manifests and suggested to add the "PDE Partial" facet to each module. After configuring `File | Settings | Languages & Frameworks | Eclipse Target`, adding _C:\Users\alexa\.p2\pool_ as am Eclipse target and plugins location, it took a while to index the directory, but now all dependencies seem to be resolved correctly and I can compile from IDEA. I.e., it might be possible to work like this and start refactoring code, still leaving the actual (published) builds up to Tycho, of course. I am quite busy and do not have much time to try today, but I hope that in the next few days I might have a free hour or two.

Sorry, but I have no intention to sponsor you, because I am working for free in OSS projects too. So unless someone sponsors my work for AspectJ, AJDT, Spock etc. first and I have some funds to pass through to you, this is not going to happen.

**Reply by laeubi** [reply]
> That the dynamic model is necessary in order to resolve all Tycho dependencies in the first place and that IDEA might not understand P2 repositories, probably does not make it a Tycho integration problem either, but IDEA's Maven support is just bad.

That is not required. You can declare all dependencies in the maven `pom.xml` and simply enable Tycho to use the pom dependencies instead of p2 sites, but most users don't like that and instead want Tycho to do this automatic :-)

> I don't know why I never found it before

Its quite new in Tycho 2.7.x

> IDEA immediately recognised the manifests and suggested to add the "PDE Partial" facet to each module.

That's how it works (most of the time) in eclipse as well, eclipse don't reads the tycho configuration (as mentioned before) it is the other way round that Tycho reads what Eclipse/PDE produces.

> I have some funds to pass through to you, this is not going to happen

No problem at all, sadly most OSS projects suffer from suitable funding.

**Reply by kriegaex** [reply]
> > I don't know why I never found it before
> 
> Its quite new in Tycho 2.7.x

That sentence was about the IDEA plugin, not about Tycho. I moved it to a separate paragraph, so maybe now it is easier to recognise even when reading quickly.
